File: server/cmd/services/games/behavior/games.go
package behavior

import (
	"github.com/google/uuid"
	t "github.com/jsbento/chess-server-v4/cmd/services/games/types"
)

func (s *Store) CreateGame(req *t.CreateGameReq) (*t.Game, error) {
	if err := req.Validate(); err != nil {
		return nil, err
	}

	game := &t.Game{
		Id:       uuid.New().String(),
		PlayerId: req.PlayerId,
		Moves:    req.Moves,
		Result:   req.Result,
	}
	if err := s.pg.GetDB().Create(game).Error; err != nil {
		return nil, err
	}
	return game, nil
}

func (s *Store) GetGame(id string) (*t.Game, error) {
	game := &t.Game{}
	if err := s.pg.GetDB().Model(&t.Game{}).Where("id = ?", id).First(game).Error; err != nil {
		return nil, err
	}
	return game, nil
}

func (s *Store) GetUserGames(playerId string) ([]*t.Game, error) {
	games := []*t.Game{}
	if err := s.pg.GetDB().Model(&t.Game{}).Where("player_id = ?", playerId).Find(&games).Error; err != nil {
		return nil, err
	}
	return games, nil
}

File: chess-ts-client/src/behavior/games/games.ts
import { post, get } from '@utils'

import type {
  SaveGameReq,
  Game,
} from '@types'

export const saveGame = async (req: SaveGameReq): Promise<Game | null> => {
  try {
    const resp = await post<SaveGameReq, Game>('/games', req)
    return resp ?? null
  } catch (err) {
    console.log(err)
    return null
  }
}

export const getUserGames = async (): Promise<Game[]> => {
  try {
    const resp = await get<null, Game[]>('/games/user')
    return resp ?? []
  } catch (err) {
    console.log(err)
    return []
  }
}